### Groupby() - 그룹화 처리
- 주제 : 타이타닉에서 선실에 등급에 따른 생존 비율을 분석

In [1]:
import pandas as pd

In [2]:
file_path = '../data/titanic.csv'

In [3]:
data_df = pd.read_csv(file_path)

In [4]:
data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [6]:
class_surv_df = data_df.loc[:, ['survived', 'pclass']]

In [9]:
class_surv_df.isna().sum()

survived    0
pclass      0
dtype: int64

In [11]:
class_surv_df.duplicated().sum()    # 의미 없잖아...

np.int64(885)

In [12]:
group_obj = class_surv_df.groupby('pclass')

In [14]:
group_obj.groups.keys()

dict_keys([1, 2, 3])

In [15]:
group_obj.describe()

survived                                             
          count      mean       std  min  25%  50%  75%  max
pclass                                                      
1         216.0  0.629630  0.484026  0.0  0.0  1.0  1.0  1.0
2         184.0  0.472826  0.500623  0.0  0.0  0.0  1.0  1.0
3         491.0  0.242363  0.428949  0.0  0.0  0.0  0.0  1.0

In [17]:
group_obj.sum(), group_obj.count()

(        survived
 pclass          
 1            136
 2             87
 3            119,
         survived
 pclass          
 1            216
 2            184
 3            491)

In [20]:
class_alive_rate = (group_obj.sum() / group_obj.count())*100

In [21]:
class_alive_rate

survived
pclass           
1       62.962963
2       47.282609
3       24.236253

- class별 승객수는 1등급 객실이 216명, 2등급 객실이 184명, 3등급 객실이 491명임.
- class별 생존자는 1등급 객실이 136명, 2등급 객실이 87명, 3등급 객실이 119임.

    - class별 생존율은 1등급 객실이 62.96%, 2등급 객실이 47.28%, 3등급 객실이 24.24%로 결과가 나타남.
    - class가 낮아질수록 생존률이 낮아지는 경향이 있는 것으로 볼 수 있음.
    - class별 사분위수를 보면 1등급 객실은 50%, 75%에서 값이 존재함을 볼수 있고 2등급객실은 75%에서 값이 존재함. 이 결과를 봐도 3등급 객실의 생존률이 매우 낮음을 알 수 있음.